vpc: optimize createMonitorServiceCommand() execution

phsm opened the pull request:
### Description
This PR moves an expensive getRouterHealthChecksConfig() upper the execution stack during VPC restart. So this way it is only executed once instead of tens or hundreds times (depending on VPC tiers amount). 

This is done by introducing a `Map<String, String> routerHealthCheckConfig` argument to createMonitorServiceCommand() method that accepts pre-computed health check config.
 
Fixes #8055

<!--- Describe your changes in DETAIL - And how has behaviour functionally changed. -->

<!-- For new features, provide link to FS, dev ML discussion etc. -->
<!-- In case of bug fix, the expected and actual behaviours, steps to reproduce. -->

<!-- When "Fixes: #<id>" is specified, the issue/PR will automatically be closed when this PR gets merged -->
<!-- For addressing multiple issues/PRs, use multiple "Fixes: #<id>" -->
<!-- Fixes: # -->

<!--- ******************************************************************************* -->
<!--- NOTE: AUTOMATION USES THE DESCRIPTIONS TO SET LABELS AND PRODUCE DOCUMENTATION. -->
<!--- PLEASE PUT AN 'X' in only **ONE** box -->
<!--- ******************************************************************************* -->

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial


### Screenshots (if appropriate):


### How Has This Been Tested?
1. Prepared a test VPC: 
- 10 tiers
- redundant offering (2 VRs)
- each tier has 1 running VM in it. 
2. Performed VPC restart with clean up. Took 61 minutes to fully restart in my case
3. Applied these changes
4. Performed VPC restart with clean up. Took 9 minutes to fully restart.

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->

#### How did you try to break this feature and the system with this change?

<!-- see how your change affects other areas of the code, etc. -->


<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/main/CONTRIBUTING.md) document -->

DaanHoogland commented:
@blueorangutan package

weizhouapache commented:
@blueorangutan package

blueorangutan commented:
@weizhouapache a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 8103

harikrishna-patnala commented:
@blueorangutan test

blueorangutan commented:
@harikrishna-patnala a [SL] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

harikrishna-patnala reviewed (commented):
results seems to be very interesting @phsm

blueorangutan commented:
<b>[SF] Trillian test result (tid-8650)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 42730 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8385-t8650-kvm-centos7.zip
Smoke tests completed. 127 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_05_list_volumes_isrecursive | `Failure` | 0.03 | test_list_volumes.py
test_07_list_volumes_listall | `Failure` | 0.02 | test_list_volumes.py

sureshanaparti commented on server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java:
```diff
@@ -2350,7 +2350,7 @@ public boolean finalizeCommandsOnStart(final Commands cmds, final VirtualMachine
                 finalizeIpAssocForNetwork(cmds, router, provider, guestNetworkId, null);
                 finalizeNetworkRulesForNetwork(cmds, router, provider, guestNetworkId);
-                finalizeMonitorService(cmds, profile, router, provider, guestNetworkId, true);
+                finalizeMonitorService(cmds, profile, router, provider, guestNetworkId, true, getRouterHealthChecksConfig(router));
```
```suggestion
                finalizeMonitorService(cmds, profile, router, provider, guestNetworkId, true, routerHealthCheckConfig);
```

@phsm seems it is called in loop here, so better get the map outside loop and pass here.

phsm replied:
Good catch, updated

sureshanaparti replied:
> Good catch, updated

thanks @phsm

sureshanaparti reviewed (approved):
Code LGTM

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
@sureshanaparti a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

weizhouapache reviewed (approved):
code lgtm

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 8130

sureshanaparti commented:
@blueorangutan test

blueorangutan commented:
@sureshanaparti a [SL] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-8666)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 49046 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8385-t8666-kvm-centos7.zip
Smoke tests completed. 121 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

phsm commented:
Is there anything else that I can do for this PR to be merged? I see some tests are failed but my Java knowledge isn't that great to actually fix them so I'd appreciate any help on this.

DaanHoogland reviewed (approved):
clgtm, not sure what we need to test for this (assuming the smoke tests have tested health check functionality)

weizhouapache commented:
> So, can this PR be merged if everyone approves? :)

Not yet. 
Needs someone to manually test it

weizhouapache commented:
@phsm 
would you like this PR to be merged into 4.18 or 4.19 ?

phsm commented:
My opinion is since the fix doesn't break any API, then it would be better to have it merged to 4.18 (+4.19, master). 
We already have this fix applied on our production systems, no issues found.

weizhouapache commented:
> My opinion is since the fix doesn't break any API, then it would be better to have it merged to 4.18 (+4.19, master). We already have this fix applied on our production systems, no issues found.

@phsm  I agree with you.
I asked you is because, if you want this to be merged into 4.18, you need to rebase with 4.18 branch.
(if I change the target branch of this PR to 4.18, there are 368 commits and lots of conflicts)

DaanHoogland commented:
@phsm I think you wanted to rebase your commit onto 4.18, did you?

weizhouapache commented:
@phsm 
please try the following commands
```
git fetch apache
git checkout apache/4.18
git branch -D main
git checkout -b main
git cherry-pick e34959fe66f6078b259041bd8023a1f73dec0a3c
git cherry-pick b6b7d47089d39f10712c027826ee0de830d6d9e0
git push origin main -f
```

phsm commented:
> @phsm please try the following commands
> 
> ```
> git fetch apache
> git checkout apache/4.18
> git branch -D main
> git checkout -b main
> git cherry-pick e34959fe66f6078b259041bd8023a1f73dec0a3c
> git cherry-pick b6b7d47089d39f10712c027826ee0de830d6d9e0
> git push origin main -f
> ```

Thanks, that seems to have worked.

vishesh92 commented:
@blueorangutan package

blueorangutan commented:
Packaging result [SF]: ✔️ el7 ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 8673

vishesh92 commented:
@blueorangutan test

blueorangutan commented:
@vishesh92 a [SL] Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-9247)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 42810 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8385-t9247-kvm-centos7.zip
Smoke tests completed. 107 look OK, 3 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_10_reboot_cpvm_forced | `Error` | 6.24 | test_ssvm.py
test_08_migrate_vm | `Error` | 0.05 | test_vm_life_cycle.py
test_01_redundant_vpc_site2site_vpn | `Failure` | 590.97 | test_vpc_vpn.py
test_01_vpc_site2site_vpn_multiple_options | `Failure` | 469.91 | test_vpc_vpn.py
test_01_vpc_site2site_vpn | `Failure` | 287.80 | test_vpc_vpn.py

weizhouapache commented:
@blueorangutan test rocky8 kvm-rocky8

blueorangutan commented:
@weizhouapache a [SL] Trillian-Jenkins test job (rocky8 mgmt + kvm-rocky8) has been kicked to run smoke tests

vishesh92 commented:
@blueorangutan test rocky8 kvm-rocky8

blueorangutan commented:
@vishesh92 a [SL] Trillian-Jenkins test job (rocky8 mgmt + kvm-rocky8) has been kicked to run smoke tests

vishesh92 commented:
Tested these changes. I couldn't find any difference in performance for these changes. I followed these steps:

1. Setup 2 envs (with and without this patch)
2. Create a VPC with redundant offering
3. Add 10 networks and launch 1 VM in each
4. Restarted VPC with cleanup
5. Check the difference in created, removed in async_job table for VPCRestartCmd

phsm commented:
> Tested these changes. I couldn't find any difference in performance for these changes. I followed these steps:
> 
>     1. Setup 2 envs (with and without this patch)
> 
>     2. Create a VPC with redundant offering
> 
>     3. Add 10 networks and launch 1 VM in each
> 
>     4. Restarted VPC with cleanup
> 
>     5. Check the difference in created, removed in async_job table for VPCRestartCmd

Hi, 
As I mentioned earlier, this is only reproducible on CS installations "with history", i.e. with a large cloudstack database with a lot of both running and removed services (VMs, networks, VPCs etc).
Only then the getRouterHealthChecksConfig() takes 1-2 seconds to complete, which with combination of hundreds of executions adds a significant time to the process.

weizhouapache commented:
> Tested these changes. I couldn't find any difference in performance for these changes. I followed these steps:
> 
> 1. Setup 2 envs (with and without this patch)
> 2. Create a VPC with redundant offering
> 3. Add 10 networks and launch 1 VM in each
> 4. Restarted VPC with cleanup
> 5. Check the difference in created, removed in async_job table for VPCRestartCmd

@vishesh92 
it is probably because our database is too small.
if the platforms have been running for many years (for example, more than 5), it could make big difference.

considering this is an optimization, it looks there is no regression (in manual and smoke tests), can we say it is good to merge ? @vishesh92

vishesh92 commented:
@weizhouapache @phsm Got it. It is good to merge from my side.

For manual testing, I just did what I mentioned in the above steps and tried to connect VMs I launched.

weizhouapache commented:
> @weizhouapache @phsm Got it. It is good to merge from my side.
> 
> For manual testing, I just did what I mentioned in the above steps and tried to connect VMs I launched.

thanks @vishesh92 for the testing

I will merge it then

blueorangutan commented:
<b>[SF] Trillian test result (tid-9286)</b>
Environment: kvm-rocky8 (x2), Advanced Networking with Mgmt server r8
Total time taken: 43840 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr8385-t9286-kvm-rocky8.zip
Smoke tests completed. 110 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
